src/core/mac/sub_mac.cpp

```diff
@@ -211,17 +227,22 @@ Error SubMac::Disable(void)
 
 Error SubMac::Sleep(void)
 {
-    Error error = Get<Radio>().Sleep();
+    Error error = kErrorNone;
+
+    VerifyOrExit(ShouldHandleTransitionToSleep());
```

Comment:
Relate to this, should we consider saving this flag in `RadioSpinel` for RCP reset recovery (to set it back to same value on reset)?

Reply:
Good point, I've added this flag to the `RadioSpinel` restoration.

One of the ways adding this feature saves power is by avoiding unneeded API calls serialization. If we call `Sleep()` each time just in case, we are going against the purpose of the PR.

I'd rather leave the responsibility to the radio platform implementation to properly handle all idle periods.

Reply:
I just realized something. What about the data poll timeout?
- After data poll tx on sleepy with parent indicating that there is pending frame, we enter rx.
- What if we don't rx any frame? (parent misses and delays before sending)
- Today, the timer in `Mac` handles this and ensures we stop `kOperationWaitingForData` which will trigger us to go idle mode and would then call `Sleep()`. 
- I think with `VerifyOrExit(ShouldHandleTransitionToSleep());` we may end up staing in rx.

I feel may be better to still call `Sleep()` (will add extra exchange with RCP but will be safer). There may be new  new use-cases that come in future and add new code paths (e.g. there are discussions on sleep-to-sleepy model).

Reply:
As I understand, it is a rare case that we let a device running host+RCP mode be a SED. Is it a common case?

Reply:
@Irving-cl some of our platforms have multiple cores in a single SoC, and for some configurations OT and radio driver belong to different cores, thus serialization of radio APIs is required.

@abtink  I'll perform controlled measurements about the power consumption impact of always calling `Sleep()` versus using `ShouldHandleTransitionToSleep()`. If the impact is low, I'd agree with your proposal. Otherwise, a boolean could be added to the `SubMac::Sleep()` method to force sleep in special cases.

Reply:
Oh that's what I think. So the 'serialization' you mean is about the SoC instructions instead of spinel right?

Make sense to me to perform a measurement to see the impact.

Reply:
Here are the measurements: increment in total charge during 1 minute polling with 1 second poll period and charge of a single data request operation, compared for a platform with IPC vs. single core SoC.

| Platform | 60s energy increment | Data poll energy increment |
| :---: |  :---: |  :---: |
| Multi-core | 6.1% | 7,4% |
| Single-core | 0.5% | 0.6% |

It seems that final product users would benefit from the extra battery duration optimization if `Sleep()` is called only when really needed.

Reply:
> I just realized something. What about the data poll timeout?

I'm now thinking that there is probably no need to add extra code complexity for this corner case, since:
- It only affects to platforms that support RxOnWhenIdle mode.
- Only when the device sends 5 data polls for which ACK with FP=1 is received, but no frame is actually received (harsh radio conditions?).
- After this, the device will be with the radio on, but only until any valid frame is received or any frame is transmitted (i.e., next data poll), moment from which radio will be turned off again.

I've played with extra `aForce` parameter in `SubMac::Sleep()`, but due to the previous reasons I believe it's not needed.

Reply:
I agree the idea of `aForce` in `Sleep()` is not appealing.

I would suggest updating the definition/documention of "idle mode" for `RxOnWhenIld` API such that there is a timeout for rx after which radio platform must put radio to sleep if no frame was received.

Reply:
That's a good compromise.

Just updated the doc and I've also added a missing idle period definition:

>  * - Finalization of a radio reception window scheduled with `otPlatRadioReceiveAt`.